### ISSUE TITLE
DASK AssertionError: data in tcp://127.0.0.1:46353 was moved.

### ISSUE THREAD

**Issue by sh1ng** [dmlc/xgboost#4987]
2GPU

Test case

https://github.com/h2oai/h2o4gpu/blob/e7ba6a43034781633630f2acbc2e9c7e75d2bbf7/tests/python/open_data/gbm/test_xgboost.py#L77-L140

Log
```
[2019-10-25T13:30:09.251Z] ________________________ test_xgboost_covtype_multi_gpu ________________________

[2019-10-25T13:30:09.251Z] [gw0] linux -- Python 3.6.4 /opt/h2oai/h2o4gpu/python/bin/python

[2019-10-25T13:30:09.251Z] 

[2019-10-25T13:30:09.251Z] self = <CallInfo when='call' exception: data in tcp://127.0.0.1:46353 was moved.>

[2019-10-25T13:30:09.251Z] func = <function call_runtest_hook.<locals>.<lambda> at 0x7fff90b6b378>

[2019-10-25T13:30:09.251Z] when = 'call', treat_keyboard_interrupt_as_exception = False

[2019-10-25T13:30:09.251Z] 

[2019-10-25T13:30:09.251Z]     def __init__(self, func, when, treat_keyboard_interrupt_as_exception=False):

[2019-10-25T13:30:09.251Z]         #: context of invocation: one of "setup", "call",

[2019-10-25T13:30:09.251Z]         #: "teardown", "memocollect"

[2019-10-25T13:30:09.251Z]         self.when = when

[2019-10-25T13:30:09.251Z]         self.start = time()

[2019-10-25T13:30:09.251Z]         try:

[2019-10-25T13:30:09.251Z] >           self.result = func()

[2019-10-25T13:30:09.251Z] 

[2019-10-25T13:30:09.251Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/_pytest/runner.py:212: 

[2019-10-25T13:30:09.251Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.251Z] 

[2019-10-25T13:30:09.251Z] >       lambda: ihook(item=item, **kwds),

[2019-10-25T13:30:09.251Z]         when=when,

[2019-10-25T13:30:09.251Z]         treat_keyboard_interrupt_as_exception=item.config.getvalue("usepdb"),

[2019-10-25T13:30:09.251Z]     )

[2019-10-25T13:30:09.251Z] 

[2019-10-25T13:30:09.251Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/_pytest/runner.py:194: 

[2019-10-25T13:30:09.251Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.251Z] 

[2019-10-25T13:30:09.251Z] self = <_HookCaller 'pytest_runtest_call'>, args = ()

[2019-10-25T13:30:09.251Z] kwargs = {'item': <Function 'test_xgboost_covtype_multi_gpu'>}

[2019-10-25T13:30:09.251Z] notincall = set()

[2019-10-25T13:30:09.251Z] 

[2019-10-25T13:30:09.251Z]     def __call__(self, *args, **kwargs):

[2019-10-25T13:30:09.251Z]         if args:

[2019-10-25T13:30:09.251Z]             raise TypeError("hook calling supports only keyword arguments")

[2019-10-25T13:30:09.251Z]         assert not self.is_historic()

[2019-10-25T13:30:09.251Z]         if self.spec and self.spec.argnames:

[2019-10-25T13:30:09.251Z]             notincall = (

[2019-10-25T13:30:09.251Z]                 set(self.spec.argnames) - set(["__multicall__"]) - set(kwargs.keys())

[2019-10-25T13:30:09.251Z]             )

[2019-10-25T13:30:09.251Z]             if notincall:

[2019-10-25T13:30:09.251Z]                 warnings.warn(

[2019-10-25T13:30:09.251Z]                     "Argument(s) {} which are declared in the hookspec "

[2019-10-25T13:30:09.251Z]                     "can not be found in this hook call".format(tuple(notincall)),

[2019-10-25T13:30:09.251Z]                     stacklevel=2,

[2019-10-25T13:30:09.251Z]                 )

[2019-10-25T13:30:09.251Z] >       return self._hookexec(self, self.get_hookimpls(), kwargs)

[2019-10-25T13:30:09.251Z] 

[2019-10-25T13:30:09.251Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/pluggy/hooks.py:286: 

[2019-10-25T13:30:09.251Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.251Z] 

[2019-10-25T13:30:09.251Z] self = <_pytest.config.PytestPluginManager object at 0x7fffbb26d7b8>

[2019-10-25T13:30:09.251Z] hook = <_HookCaller 'pytest_runtest_call'>

[2019-10-25T13:30:09.251Z] methods = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '/opt/h2oai/h2o4gpu/python/lib/python3.6/site-pa...ffbabbacc0>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0x7fffba055518>>]

[2019-10-25T13:30:09.251Z] kwargs = {'item': <Function 'test_xgboost_covtype_multi_gpu'>}

[2019-10-25T13:30:09.251Z] 

[2019-10-25T13:30:09.251Z]     def _hookexec(self, hook, methods, kwargs):

[2019-10-25T13:30:09.251Z]         # called from all hookcaller instances.

[2019-10-25T13:30:09.251Z]         # enable_tracing will set its own wrapping function at self._inner_hookexec

[2019-10-25T13:30:09.251Z] >       return self._inner_hookexec(hook, methods, kwargs)

[2019-10-25T13:30:09.251Z] 

[2019-10-25T13:30:09.251Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/pluggy/manager.py:92: 

[2019-10-25T13:30:09.251Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.251Z] 

[2019-10-25T13:30:09.251Z] hook = <_HookCaller 'pytest_runtest_call'>

[2019-10-25T13:30:09.251Z] methods = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '/opt/h2oai/h2o4gpu/python/lib/python3.6/site-pa...ffbabbacc0>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0x7fffba055518>>]

[2019-10-25T13:30:09.251Z] kwargs = {'item': <Function 'test_xgboost_covtype_multi_gpu'>}

[2019-10-25T13:30:09.251Z] 

[2019-10-25T13:30:09.251Z]     self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(

[2019-10-25T13:30:09.251Z]         methods,

[2019-10-25T13:30:09.251Z]         kwargs,

[2019-10-25T13:30:09.251Z] >       firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,

[2019-10-25T13:30:09.251Z]     )

[2019-10-25T13:30:09.251Z] 

[2019-10-25T13:30:09.252Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/pluggy/manager.py:86: 

[2019-10-25T13:30:09.252Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z] item = <Function 'test_xgboost_covtype_multi_gpu'>

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z]     def pytest_runtest_call(item):

[2019-10-25T13:30:09.252Z]         _update_current_test_var(item, "call")

[2019-10-25T13:30:09.252Z]         sys.last_type, sys.last_value, sys.last_traceback = (None, None, None)

[2019-10-25T13:30:09.252Z]         try:

[2019-10-25T13:30:09.252Z] >           item.runtest()

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/_pytest/runner.py:122: 

[2019-10-25T13:30:09.252Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z] self = <Function 'test_xgboost_covtype_multi_gpu'>

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z]     def runtest(self):

[2019-10-25T13:30:09.252Z]         """ execute the underlying test function. """

[2019-10-25T13:30:09.252Z] >       self.ihook.pytest_pyfunc_call(pyfuncitem=self)

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/_pytest/python.py:1438: 

[2019-10-25T13:30:09.252Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z] self = <_HookCaller 'pytest_pyfunc_call'>, args = ()

[2019-10-25T13:30:09.252Z] kwargs = {'pyfuncitem': <Function 'test_xgboost_covtype_multi_gpu'>}

[2019-10-25T13:30:09.252Z] notincall = set()

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z]     def __call__(self, *args, **kwargs):

[2019-10-25T13:30:09.252Z]         if args:

[2019-10-25T13:30:09.252Z]             raise TypeError("hook calling supports only keyword arguments")

[2019-10-25T13:30:09.252Z]         assert not self.is_historic()

[2019-10-25T13:30:09.252Z]         if self.spec and self.spec.argnames:

[2019-10-25T13:30:09.252Z]             notincall = (

[2019-10-25T13:30:09.252Z]                 set(self.spec.argnames) - set(["__multicall__"]) - set(kwargs.keys())

[2019-10-25T13:30:09.252Z]             )

[2019-10-25T13:30:09.252Z]             if notincall:

[2019-10-25T13:30:09.252Z]                 warnings.warn(

[2019-10-25T13:30:09.252Z]                     "Argument(s) {} which are declared in the hookspec "

[2019-10-25T13:30:09.252Z]                     "can not be found in this hook call".format(tuple(notincall)),

[2019-10-25T13:30:09.252Z]                     stacklevel=2,

[2019-10-25T13:30:09.252Z]                 )

[2019-10-25T13:30:09.252Z] >       return self._hookexec(self, self.get_hookimpls(), kwargs)

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/pluggy/hooks.py:286: 

[2019-10-25T13:30:09.252Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z] self = <_pytest.config.PytestPluginManager object at 0x7fffbb26d7b8>

[2019-10-25T13:30:09.252Z] hook = <_HookCaller 'pytest_pyfunc_call'>

[2019-10-25T13:30:09.252Z] methods = [<HookImpl plugin_name='python', plugin=<module '_pytest.python' from '/opt/h2oai/h2o4gpu/python/lib/python3.6/site-pa..., plugin=<module '_pytest.skipping' from '/opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/_pytest/skipping.py'>>]

[2019-10-25T13:30:09.252Z] kwargs = {'pyfuncitem': <Function 'test_xgboost_covtype_multi_gpu'>}

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z]     def _hookexec(self, hook, methods, kwargs):

[2019-10-25T13:30:09.252Z]         # called from all hookcaller instances.

[2019-10-25T13:30:09.252Z]         # enable_tracing will set its own wrapping function at self._inner_hookexec

[2019-10-25T13:30:09.252Z] >       return self._inner_hookexec(hook, methods, kwargs)

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/pluggy/manager.py:92: 

[2019-10-25T13:30:09.252Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z] hook = <_HookCaller 'pytest_pyfunc_call'>

[2019-10-25T13:30:09.252Z] methods = [<HookImpl plugin_name='python', plugin=<module '_pytest.python' from '/opt/h2oai/h2o4gpu/python/lib/python3.6/site-pa..., plugin=<module '_pytest.skipping' from '/opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/_pytest/skipping.py'>>]

[2019-10-25T13:30:09.252Z] kwargs = {'pyfuncitem': <Function 'test_xgboost_covtype_multi_gpu'>}

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z]     self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(

[2019-10-25T13:30:09.252Z]         methods,

[2019-10-25T13:30:09.252Z]         kwargs,

[2019-10-25T13:30:09.252Z] >       firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,

[2019-10-25T13:30:09.252Z]     )

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/pluggy/manager.py:86: 

[2019-10-25T13:30:09.252Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z] pyfuncitem = <Function 'test_xgboost_covtype_multi_gpu'>

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z]     @hookimpl(trylast=True)

[2019-10-25T13:30:09.252Z]     def pytest_pyfunc_call(pyfuncitem):

[2019-10-25T13:30:09.252Z]         testfunction = pyfuncitem.obj

[2019-10-25T13:30:09.252Z]         if pyfuncitem._isyieldedfunction():

[2019-10-25T13:30:09.252Z]             testfunction(*pyfuncitem._args)

[2019-10-25T13:30:09.252Z]         else:

[2019-10-25T13:30:09.252Z]             funcargs = pyfuncitem.funcargs

[2019-10-25T13:30:09.252Z]             testargs = {}

[2019-10-25T13:30:09.252Z]             for arg in pyfuncitem._fixtureinfo.argnames:

[2019-10-25T13:30:09.252Z]                 testargs[arg] = funcargs[arg]

[2019-10-25T13:30:09.252Z] >           testfunction(**testargs)

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/_pytest/python.py:166: 

[2019-10-25T13:30:09.252Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z]     @pytest.mark.multi_gpu

[2019-10-25T13:30:09.252Z]     def test_xgboost_covtype_multi_gpu():

[2019-10-25T13:30:09.252Z]         import xgboost as xgb

[2019-10-25T13:30:09.252Z]         import numpy as np

[2019-10-25T13:30:09.252Z]         from sklearn.datasets import fetch_covtype

[2019-10-25T13:30:09.252Z]         from sklearn.model_selection import train_test_split

[2019-10-25T13:30:09.252Z]         import time

[2019-10-25T13:30:09.252Z]         from dask_cuda import LocalCUDACluster

[2019-10-25T13:30:09.252Z]         from dask.distributed import Client

[2019-10-25T13:30:09.252Z]         from dask import array as da

[2019-10-25T13:30:09.252Z]         import xgboost as xgb

[2019-10-25T13:30:09.252Z]         from xgboost.dask import DaskDMatrix

[2019-10-25T13:30:09.252Z]         from dask import array as da

[2019-10-25T13:30:09.252Z]     

[2019-10-25T13:30:09.252Z]         # Fetch dataset using sklearn

[2019-10-25T13:30:09.252Z]         cov = fetch_covtype()

[2019-10-25T13:30:09.252Z]         X = cov.data

[2019-10-25T13:30:09.252Z]         y = cov.target

[2019-10-25T13:30:09.252Z]     

[2019-10-25T13:30:09.252Z]         print(X.shape, y.shape)

[2019-10-25T13:30:09.252Z]     

[2019-10-25T13:30:09.252Z]         # Create 0.75/0.25 train/test split

[2019-10-25T13:30:09.252Z]         X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, train_size=0.75,

[2019-10-25T13:30:09.252Z]                                                             random_state=42)

[2019-10-25T13:30:09.252Z]     

[2019-10-25T13:30:09.252Z]         # Specify sufficient boosting iterations to reach a minimum

[2019-10-25T13:30:09.252Z]         num_round = 10

[2019-10-25T13:30:09.252Z]     

[2019-10-25T13:30:09.252Z]         # Leave most parameters as default

[2019-10-25T13:30:09.252Z]         param = {'objective': 'multi:softmax',  # Specify multiclass classification

[2019-10-25T13:30:09.252Z]                  'num_class': 8,  # Number of possible output classes

[2019-10-25T13:30:09.252Z]                  'tree_method': 'gpu_hist',  # Use GPU accelerated algorithm

[2019-10-25T13:30:09.252Z]                  }

[2019-10-25T13:30:09.252Z]     

[2019-10-25T13:30:09.252Z]         from h2o4gpu.util.gpu import device_count

[2019-10-25T13:30:09.252Z]         n_gpus, devices = device_count(-1)

[2019-10-25T13:30:09.252Z]     

[2019-10-25T13:30:09.252Z]         with LocalCUDACluster(n_workers=n_gpus, threads_per_worker=1) as cluster:

[2019-10-25T13:30:09.252Z]             with Client(cluster) as client:

[2019-10-25T13:30:09.252Z]                 # Convert input data from numpy to XGBoost format

[2019-10-25T13:30:09.252Z]                 partition_size = 100000

[2019-10-25T13:30:09.252Z]                 dask_X_train = da.from_array(X_train, partition_size)

[2019-10-25T13:30:09.252Z]                 dask_y_label = da.from_array(y_train, partition_size)

[2019-10-25T13:30:09.252Z]                 dtrain = DaskDMatrix(

[2019-10-25T13:30:09.252Z]                     client=client, data=dask_X_train, label=dask_y_label)

[2019-10-25T13:30:09.252Z]                 dtest = DaskDMatrix(client=client, data=da.from_array(

[2019-10-25T13:30:09.252Z]                     X_test, partition_size), label=da.from_array(y_test, partition_size))

[2019-10-25T13:30:09.252Z]     

[2019-10-25T13:30:09.252Z]                 gpu_res = {}  # Store accuracy result

[2019-10-25T13:30:09.252Z]                 tmp = time.time()

[2019-10-25T13:30:09.252Z]                 # Train model

[2019-10-25T13:30:09.252Z]                 xgb.dask.train(client, param, dtrain, num_boost_round=num_round, evals=[

[2019-10-25T13:30:09.252Z] >                   (dtest, 'test')])

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z] tests/python/open_data/gbm/test_xgboost.py:128: 

[2019-10-25T13:30:09.252Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z] client = <Client: not connected>

[2019-10-25T13:30:09.252Z] params = {'num_class': 8, 'objective': 'multi:softmax', 'tree_method': 'gpu_hist'}

[2019-10-25T13:30:09.252Z] dtrain = <xgboost.dask.DaskDMatrix object at 0x7fffad769e80>

[2019-10-25T13:30:09.252Z] evals = [(<xgboost.dask.DaskDMatrix object at 0x7fffad769cc0>, 'test')]

[2019-10-25T13:30:09.252Z] args = (), kwargs = {'num_boost_round': 10}

[2019-10-25T13:30:09.252Z] worker_map = defaultdict(<class 'list'>, {'tcp://127.0.0.1:35149': [<Future: status: cancelled, type: tuple, key: tuple-4bb2a22b-fd...9-4ff0-9d77-0945e51d31a9>, <Future: status: cancelled, type: tuple, key: tuple-79909886-6adc-4c30-9ca0-f2a3150e7083>]})

[2019-10-25T13:30:09.252Z] dispatched_train = <function train.<locals>.dispatched_train at 0x7fff5e80d400>

[2019-10-25T13:30:09.252Z] futures = [<Future: status: cancelled, key: dispatched_train-aeec3bd0-915c-46ba-9098-9dcf8d816571-0>, <Future: status: cancelled, key: dispatched_train-aeec3bd0-915c-46ba-9098-9dcf8d816571-1>]

[2019-10-25T13:30:09.252Z] 

[2019-10-25T13:30:09.252Z]     def train(client, params, dtrain, *args, evals=(), **kwargs):

[2019-10-25T13:30:09.252Z]         '''Train XGBoost model.

[2019-10-25T13:30:09.252Z]     

[2019-10-25T13:30:09.252Z]         Parameters

[2019-10-25T13:30:09.252Z]         ----------

[2019-10-25T13:30:09.252Z]         client: dask.distributed.Client

[2019-10-25T13:30:09.252Z]           Specify the dask client used for training.  Use default client

[2019-10-25T13:30:09.252Z]           returned from dask if it's set to None.

[2019-10-25T13:30:09.252Z]     

[2019-10-25T13:30:09.252Z]         Other parameters are the same as `xgboost.train` except for `evals_result`,

[2019-10-25T13:30:09.252Z]         which is returned as part of function return value instead of argument.

[2019-10-25T13:30:09.252Z]     

[2019-10-25T13:30:09.252Z]         Returns

[2019-10-25T13:30:09.252Z]         -------

[2019-10-25T13:30:09.252Z]         results: dict

[2019-10-25T13:30:09.252Z]           A dictionary containing trained booster and evaluation history.

[2019-10-25T13:30:09.252Z]          `history` field is the same as `eval_result` from `xgboost.train`.

[2019-10-25T13:30:09.252Z]     

[2019-10-25T13:30:09.252Z]           .. code-block:: python

[2019-10-25T13:30:09.252Z]     

[2019-10-25T13:30:09.252Z]             {'booster': xgboost.Booster,

[2019-10-25T13:30:09.252Z]              'history': {'train': {'logloss': ['0.48253', '0.35953']},

[2019-10-25T13:30:09.252Z]                          'eval': {'logloss': ['0.480385', '0.357756']}}}

[2019-10-25T13:30:09.252Z]     

[2019-10-25T13:30:09.252Z]         '''

[2019-10-25T13:30:09.252Z]         _assert_dask_installed()

[2019-10-25T13:30:09.252Z]         if platform.system() == 'Windows':

[2019-10-25T13:30:09.252Z]             msg = 'Windows is not officially supported for dask/xgboost,'

[2019-10-25T13:30:09.252Z]             msg += ' contribution are welcomed.'

[2019-10-25T13:30:09.252Z]             logging.warning(msg)

[2019-10-25T13:30:09.252Z]     

[2019-10-25T13:30:09.252Z]         if 'evals_result' in kwargs.keys():

[2019-10-25T13:30:09.252Z]             raise ValueError(

[2019-10-25T13:30:09.252Z]                 'evals_result is not supported in dask interface.',

[2019-10-25T13:30:09.252Z]                 'The evaluation history is returned as result of training.')

[2019-10-25T13:30:09.252Z]     

[2019-10-25T13:30:09.252Z]         client = _xgb_get_client(client)

[2019-10-25T13:30:09.252Z]     

[2019-10-25T13:30:09.252Z]         worker_map = dtrain.worker_map

[2019-10-25T13:30:09.252Z]         rabit_args = _get_rabit_args(worker_map, client)

[2019-10-25T13:30:09.252Z]     

[2019-10-25T13:30:09.252Z]         def dispatched_train(worker_id):

[2019-10-25T13:30:09.253Z]             '''Perform training on worker.'''

[2019-10-25T13:30:09.253Z]             logging.info('Training on %d', worker_id)

[2019-10-25T13:30:09.253Z]             worker = distributed_get_worker()

[2019-10-25T13:30:09.253Z]             local_dtrain = dtrain.get_worker_data(worker)

[2019-10-25T13:30:09.253Z]     

[2019-10-25T13:30:09.253Z]             local_evals = []

[2019-10-25T13:30:09.253Z]             if evals:

[2019-10-25T13:30:09.253Z]                 for mat, name in evals:

[2019-10-25T13:30:09.253Z]                     local_mat = mat.get_worker_data(worker)

[2019-10-25T13:30:09.253Z]                     local_evals.append((local_mat, name))

[2019-10-25T13:30:09.253Z]     

[2019-10-25T13:30:09.253Z]             with RabitContext(rabit_args):

[2019-10-25T13:30:09.253Z]                 local_history = {}

[2019-10-25T13:30:09.253Z]                 local_param = params.copy()  # just to be consistent

[2019-10-25T13:30:09.253Z]                 bst = worker_train(params=local_param,

[2019-10-25T13:30:09.253Z]                                    dtrain=local_dtrain,

[2019-10-25T13:30:09.253Z]                                    *args,

[2019-10-25T13:30:09.253Z]                                    evals_result=local_history,

[2019-10-25T13:30:09.253Z]                                    evals=local_evals,

[2019-10-25T13:30:09.253Z]                                    **kwargs)

[2019-10-25T13:30:09.253Z]                 ret = {'booster': bst, 'history': local_history}

[2019-10-25T13:30:09.253Z]                 if rabit.get_rank() != 0:

[2019-10-25T13:30:09.253Z]                     ret = None

[2019-10-25T13:30:09.253Z]                 return ret

[2019-10-25T13:30:09.253Z]     

[2019-10-25T13:30:09.253Z]         futures = client.map(dispatched_train,

[2019-10-25T13:30:09.253Z]                              range(len(worker_map)),

[2019-10-25T13:30:09.253Z]                              pure=False,

[2019-10-25T13:30:09.253Z]                              workers=list(worker_map.keys()))

[2019-10-25T13:30:09.253Z] >       results = client.gather(futures)

[2019-10-25T13:30:09.253Z] 

[2019-10-25T13:30:09.253Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/xgboost/dask.py:391: 

[2019-10-25T13:30:09.253Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.253Z] 

[2019-10-25T13:30:09.253Z] self = <Client: not connected>

[2019-10-25T13:30:09.253Z] futures = [<Future: status: cancelled, key: dispatched_train-aeec3bd0-915c-46ba-9098-9dcf8d816571-0>, <Future: status: cancelled, key: dispatched_train-aeec3bd0-915c-46ba-9098-9dcf8d816571-1>]

[2019-10-25T13:30:09.253Z] errors = 'raise', maxsize = 0, direct = None, asynchronous = None

[2019-10-25T13:30:09.253Z] 

[2019-10-25T13:30:09.253Z]     def gather(

[2019-10-25T13:30:09.253Z]         self, futures, errors="raise", maxsize=0, direct=None, asynchronous=None

[2019-10-25T13:30:09.253Z]     ):

[2019-10-25T13:30:09.253Z]         """ Gather futures from distributed memory

[2019-10-25T13:30:09.253Z]     

[2019-10-25T13:30:09.253Z]         Accepts a future, nested container of futures, iterator, or queue.

[2019-10-25T13:30:09.253Z]         The return type will match the input type.

[2019-10-25T13:30:09.253Z]     

[2019-10-25T13:30:09.253Z]         Parameters

[2019-10-25T13:30:09.253Z]         ----------

[2019-10-25T13:30:09.253Z]         futures: Collection of futures

[2019-10-25T13:30:09.253Z]             This can be a possibly nested collection of Future objects.

[2019-10-25T13:30:09.253Z]             Collections can be lists, sets, iterators, queues or dictionaries

[2019-10-25T13:30:09.253Z]         errors: string

[2019-10-25T13:30:09.253Z]             Either 'raise' or 'skip' if we should raise if a future has erred

[2019-10-25T13:30:09.253Z]             or skip its inclusion in the output collection

[2019-10-25T13:30:09.253Z]         direct: boolean

[2019-10-25T13:30:09.253Z]             Whether or not to connect directly to the workers, or to ask

[2019-10-25T13:30:09.253Z]             the scheduler to serve as intermediary.  This can also be set when

[2019-10-25T13:30:09.253Z]             creating the Client.

[2019-10-25T13:30:09.253Z]         maxsize: int

[2019-10-25T13:30:09.253Z]             If the input is a queue then this produces an output queue with a

[2019-10-25T13:30:09.253Z]             maximum size.

[2019-10-25T13:30:09.253Z]     

[2019-10-25T13:30:09.253Z]         Returns

[2019-10-25T13:30:09.253Z]         -------

[2019-10-25T13:30:09.253Z]         results: a collection of the same type as the input, but now with

[2019-10-25T13:30:09.253Z]         gathered results rather than futures

[2019-10-25T13:30:09.253Z]     

[2019-10-25T13:30:09.253Z]         Examples

[2019-10-25T13:30:09.253Z]         --------

[2019-10-25T13:30:09.253Z]         >>> from operator import add  # doctest: +SKIP

[2019-10-25T13:30:09.253Z]         >>> c = Client('127.0.0.1:8787')  # doctest: +SKIP

[2019-10-25T13:30:09.253Z]         >>> x = c.submit(add, 1, 2)  # doctest: +SKIP

[2019-10-25T13:30:09.253Z]         >>> c.gather(x)  # doctest: +SKIP

[2019-10-25T13:30:09.253Z]         3

[2019-10-25T13:30:09.253Z]         >>> c.gather([x, [x], x])  # support lists and dicts # doctest: +SKIP

[2019-10-25T13:30:09.253Z]         [3, [3], 3]

[2019-10-25T13:30:09.253Z]     

[2019-10-25T13:30:09.253Z]         >>> seq = c.gather(iter([x, x]))  # support iterators # doctest: +SKIP

[2019-10-25T13:30:09.253Z]         >>> next(seq)  # doctest: +SKIP

[2019-10-25T13:30:09.253Z]         3

[2019-10-25T13:30:09.253Z]     

[2019-10-25T13:30:09.253Z]         See Also

[2019-10-25T13:30:09.253Z]         --------

[2019-10-25T13:30:09.253Z]         Client.scatter: Send data out to cluster

[2019-10-25T13:30:09.253Z]         """

[2019-10-25T13:30:09.253Z]         if isqueue(futures):

[2019-10-25T13:30:09.253Z]             qout = pyQueue(maxsize=maxsize)

[2019-10-25T13:30:09.253Z]             t = threading.Thread(

[2019-10-25T13:30:09.253Z]                 target=self._threaded_gather,

[2019-10-25T13:30:09.253Z]                 name="Threaded gather()",

[2019-10-25T13:30:09.253Z]                 args=(futures, qout),

[2019-10-25T13:30:09.253Z]                 kwargs={"errors": errors, "direct": direct},

[2019-10-25T13:30:09.253Z]             )

[2019-10-25T13:30:09.253Z]             t.daemon = True

[2019-10-25T13:30:09.253Z]             t.start()

[2019-10-25T13:30:09.253Z]             return qout

[2019-10-25T13:30:09.253Z]         elif isinstance(futures, Iterator):

[2019-10-25T13:30:09.253Z]             return (self.gather(f, errors=errors, direct=direct) for f in futures)

[2019-10-25T13:30:09.253Z]         else:

[2019-10-25T13:30:09.253Z]             if hasattr(thread_state, "execution_state"):  # within worker task

[2019-10-25T13:30:09.253Z]                 local_worker = thread_state.execution_state["worker"]

[2019-10-25T13:30:09.253Z]             else:

[2019-10-25T13:30:09.253Z]                 local_worker = None

[2019-10-25T13:30:09.253Z]             return self.sync(

[2019-10-25T13:30:09.253Z]                 self._gather,

[2019-10-25T13:30:09.253Z]                 futures,

[2019-10-25T13:30:09.253Z]                 errors=errors,

[2019-10-25T13:30:09.253Z]                 direct=direct,

[2019-10-25T13:30:09.253Z]                 local_worker=local_worker,

[2019-10-25T13:30:09.253Z] >               asynchronous=asynchronous,

[2019-10-25T13:30:09.253Z]             )

[2019-10-25T13:30:09.253Z] 

[2019-10-25T13:30:09.253Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/distributed/client.py:1822: 

[2019-10-25T13:30:09.253Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.253Z] 

[2019-10-25T13:30:09.253Z] self = <Client: not connected>

[2019-10-25T13:30:09.253Z] func = <bound method Client._gather of <Client: not connected>>

[2019-10-25T13:30:09.253Z] args = ([<Future: status: cancelled, key: dispatched_train-aeec3bd0-915c-46ba-9098-9dcf8d816571-0>, <Future: status: cancelled, key: dispatched_train-aeec3bd0-915c-46ba-9098-9dcf8d816571-1>],)

[2019-10-25T13:30:09.253Z] kwargs = {'direct': None, 'errors': 'raise', 'local_worker': None}

[2019-10-25T13:30:09.253Z] asynchronous = None

[2019-10-25T13:30:09.253Z] 

[2019-10-25T13:30:09.253Z]     def sync(self, func, *args, **kwargs):

[2019-10-25T13:30:09.253Z]         asynchronous = kwargs.pop("asynchronous", None)

[2019-10-25T13:30:09.253Z]         if (

[2019-10-25T13:30:09.253Z]             asynchronous

[2019-10-25T13:30:09.253Z]             or self.asynchronous

[2019-10-25T13:30:09.253Z]             or getattr(thread_state, "asynchronous", False)

[2019-10-25T13:30:09.253Z]         ):

[2019-10-25T13:30:09.253Z]             callback_timeout = kwargs.pop("callback_timeout", None)

[2019-10-25T13:30:09.253Z]             future = func(*args, **kwargs)

[2019-10-25T13:30:09.253Z]             if callback_timeout is not None:

[2019-10-25T13:30:09.253Z]                 future = gen.with_timeout(timedelta(seconds=callback_timeout), future)

[2019-10-25T13:30:09.253Z]             return future

[2019-10-25T13:30:09.253Z]         else:

[2019-10-25T13:30:09.253Z] >           return sync(self.loop, func, *args, **kwargs)

[2019-10-25T13:30:09.253Z] 

[2019-10-25T13:30:09.253Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/distributed/client.py:753: 

[2019-10-25T13:30:09.253Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.253Z] 

[2019-10-25T13:30:09.253Z] loop = <tornado.platform.asyncio.AsyncIOLoop object at 0x7fff90a9ca90>

[2019-10-25T13:30:09.253Z] func = <bound method Client._gather of <Client: not connected>>

[2019-10-25T13:30:09.253Z] args = ([<Future: status: cancelled, key: dispatched_train-aeec3bd0-915c-46ba-9098-9dcf8d816571-0>, <Future: status: cancelled, key: dispatched_train-aeec3bd0-915c-46ba-9098-9dcf8d816571-1>],)

[2019-10-25T13:30:09.253Z] kwargs = {'direct': None, 'errors': 'raise', 'local_worker': None}

[2019-10-25T13:30:09.253Z] f = <function sync.<locals>.f at 0x7fff2e5cb1e0>

[2019-10-25T13:30:09.253Z] 

[2019-10-25T13:30:09.253Z]     def sync(loop, func, *args, **kwargs):

[2019-10-25T13:30:09.253Z]         """

[2019-10-25T13:30:09.253Z]         Run coroutine in loop running in separate thread.

[2019-10-25T13:30:09.253Z]         """

[2019-10-25T13:30:09.253Z]         # Tornado's PollIOLoop doesn't raise when using closed, do it ourselves

[2019-10-25T13:30:09.253Z]         if PollIOLoop and (

[2019-10-25T13:30:09.253Z]             (isinstance(loop, PollIOLoop) and getattr(loop, "_closing", False))

[2019-10-25T13:30:09.253Z]             or (hasattr(loop, "asyncio_loop") and loop.asyncio_loop._closed)

[2019-10-25T13:30:09.253Z]         ):

[2019-10-25T13:30:09.253Z]             raise RuntimeError("IOLoop is closed")

[2019-10-25T13:30:09.253Z]         try:

[2019-10-25T13:30:09.253Z]             if loop.asyncio_loop.is_closed():  # tornado 6

[2019-10-25T13:30:09.253Z]                 raise RuntimeError("IOLoop is closed")

[2019-10-25T13:30:09.253Z]         except AttributeError:

[2019-10-25T13:30:09.253Z]             pass

[2019-10-25T13:30:09.253Z]     

[2019-10-25T13:30:09.253Z]         timeout = kwargs.pop("callback_timeout", None)

[2019-10-25T13:30:09.253Z]     

[2019-10-25T13:30:09.253Z]         e = threading.Event()

[2019-10-25T13:30:09.253Z]         main_tid = get_thread_identity()

[2019-10-25T13:30:09.253Z]         result = [None]

[2019-10-25T13:30:09.253Z]         error = [False]

[2019-10-25T13:30:09.253Z]     

[2019-10-25T13:30:09.253Z]         @gen.coroutine

[2019-10-25T13:30:09.253Z]         def f():

[2019-10-25T13:30:09.253Z]             try:

[2019-10-25T13:30:09.253Z]                 if main_tid == get_thread_identity():

[2019-10-25T13:30:09.253Z]                     raise RuntimeError("sync() called from thread of running loop")

[2019-10-25T13:30:09.253Z]                 yield gen.moment

[2019-10-25T13:30:09.253Z]                 thread_state.asynchronous = True

[2019-10-25T13:30:09.253Z]                 future = func(*args, **kwargs)

[2019-10-25T13:30:09.253Z]                 if timeout is not None:

[2019-10-25T13:30:09.253Z]                     future = gen.with_timeout(timedelta(seconds=timeout), future)

[2019-10-25T13:30:09.253Z]                 result[0] = yield future

[2019-10-25T13:30:09.253Z]             except Exception as exc:

[2019-10-25T13:30:09.253Z]                 error[0] = sys.exc_info()

[2019-10-25T13:30:09.253Z]             finally:

[2019-10-25T13:30:09.253Z]                 thread_state.asynchronous = False

[2019-10-25T13:30:09.253Z]                 e.set()

[2019-10-25T13:30:09.253Z]     

[2019-10-25T13:30:09.253Z]         loop.add_callback(f)

[2019-10-25T13:30:09.253Z]         if timeout is not None:

[2019-10-25T13:30:09.253Z]             if not e.wait(timeout):

[2019-10-25T13:30:09.253Z]                 raise gen.TimeoutError("timed out after %s s." % (timeout,))

[2019-10-25T13:30:09.253Z]         else:

[2019-10-25T13:30:09.253Z]             while not e.is_set():

[2019-10-25T13:30:09.253Z]                 e.wait(10)

[2019-10-25T13:30:09.253Z]         if error[0]:

[2019-10-25T13:30:09.253Z] >           six.reraise(*error[0])

[2019-10-25T13:30:09.253Z] 

[2019-10-25T13:30:09.253Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/distributed/utils.py:331: 

[2019-10-25T13:30:09.253Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.253Z] 

[2019-10-25T13:30:09.253Z] tp = <class 'AssertionError'>, value = None, tb = None

[2019-10-25T13:30:09.253Z] 

[2019-10-25T13:30:09.253Z]     def reraise(tp, value, tb=None):

[2019-10-25T13:30:09.253Z]         try:

[2019-10-25T13:30:09.253Z]             if value is None:

[2019-10-25T13:30:09.253Z]                 value = tp()

[2019-10-25T13:30:09.253Z]             if value.__traceback__ is not tb:

[2019-10-25T13:30:09.253Z]                 raise value.with_traceback(tb)

[2019-10-25T13:30:09.254Z] >           raise value

[2019-10-25T13:30:09.254Z] 

[2019-10-25T13:30:09.254Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/six.py:693: 

[2019-10-25T13:30:09.254Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.254Z] 

[2019-10-25T13:30:09.254Z]     @gen.coroutine

[2019-10-25T13:30:09.254Z]     def f():

[2019-10-25T13:30:09.254Z]         try:

[2019-10-25T13:30:09.254Z]             if main_tid == get_thread_identity():

[2019-10-25T13:30:09.254Z]                 raise RuntimeError("sync() called from thread of running loop")

[2019-10-25T13:30:09.254Z]             yield gen.moment

[2019-10-25T13:30:09.254Z]             thread_state.asynchronous = True

[2019-10-25T13:30:09.254Z]             future = func(*args, **kwargs)

[2019-10-25T13:30:09.254Z]             if timeout is not None:

[2019-10-25T13:30:09.254Z]                 future = gen.with_timeout(timedelta(seconds=timeout), future)

[2019-10-25T13:30:09.254Z] >           result[0] = yield future

[2019-10-25T13:30:09.254Z] 

[2019-10-25T13:30:09.254Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/distributed/utils.py:316: 

[2019-10-25T13:30:09.254Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.254Z] 

[2019-10-25T13:30:09.254Z] self = <tornado.gen.Runner object at 0x7fff5e7d0d30>

[2019-10-25T13:30:09.254Z] 

[2019-10-25T13:30:09.254Z]     def run(self) -> None:

[2019-10-25T13:30:09.254Z]         """Starts or resumes the generator, running until it reaches a

[2019-10-25T13:30:09.254Z]         yield point that is not ready.

[2019-10-25T13:30:09.254Z]         """

[2019-10-25T13:30:09.254Z]         if self.running or self.finished:

[2019-10-25T13:30:09.254Z]             return

[2019-10-25T13:30:09.254Z]         try:

[2019-10-25T13:30:09.254Z]             self.running = True

[2019-10-25T13:30:09.254Z]             while True:

[2019-10-25T13:30:09.254Z]                 future = self.future

[2019-10-25T13:30:09.254Z]                 if future is None:

[2019-10-25T13:30:09.254Z]                     raise Exception("No pending future")

[2019-10-25T13:30:09.254Z]                 if not future.done():

[2019-10-25T13:30:09.254Z]                     return

[2019-10-25T13:30:09.254Z]                 self.future = None

[2019-10-25T13:30:09.254Z]                 try:

[2019-10-25T13:30:09.254Z]                     exc_info = None

[2019-10-25T13:30:09.254Z]     

[2019-10-25T13:30:09.254Z]                     try:

[2019-10-25T13:30:09.254Z] >                       value = future.result()

[2019-10-25T13:30:09.254Z] 

[2019-10-25T13:30:09.254Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/tornado/gen.py:735: 

[2019-10-25T13:30:09.254Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.254Z] 

[2019-10-25T13:30:09.254Z] self = <tornado.gen.Runner object at 0x7fffad769630>

[2019-10-25T13:30:09.254Z] 

[2019-10-25T13:30:09.254Z]     def run(self) -> None:

[2019-10-25T13:30:09.254Z]         """Starts or resumes the generator, running until it reaches a

[2019-10-25T13:30:09.254Z]         yield point that is not ready.

[2019-10-25T13:30:09.254Z]         """

[2019-10-25T13:30:09.254Z]         if self.running or self.finished:

[2019-10-25T13:30:09.254Z]             return

[2019-10-25T13:30:09.254Z]         try:

[2019-10-25T13:30:09.254Z]             self.running = True

[2019-10-25T13:30:09.254Z]             while True:

[2019-10-25T13:30:09.254Z]                 future = self.future

[2019-10-25T13:30:09.254Z]                 if future is None:

[2019-10-25T13:30:09.254Z]                     raise Exception("No pending future")

[2019-10-25T13:30:09.254Z]                 if not future.done():

[2019-10-25T13:30:09.254Z]                     return

[2019-10-25T13:30:09.254Z]                 self.future = None

[2019-10-25T13:30:09.254Z]                 try:

[2019-10-25T13:30:09.254Z]                     exc_info = None

[2019-10-25T13:30:09.254Z]     

[2019-10-25T13:30:09.254Z]                     try:

[2019-10-25T13:30:09.254Z]                         value = future.result()

[2019-10-25T13:30:09.254Z]                     except Exception:

[2019-10-25T13:30:09.254Z]                         exc_info = sys.exc_info()

[2019-10-25T13:30:09.254Z]                     future = None

[2019-10-25T13:30:09.254Z]     

[2019-10-25T13:30:09.254Z]                     if exc_info is not None:

[2019-10-25T13:30:09.254Z]                         try:

[2019-10-25T13:30:09.254Z] >                           yielded = self.gen.throw(*exc_info)  # type: ignore

[2019-10-25T13:30:09.254Z] 

[2019-10-25T13:30:09.254Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/tornado/gen.py:742: 

[2019-10-25T13:30:09.254Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.254Z] 

[2019-10-25T13:30:09.254Z] self = <Client: not connected>

[2019-10-25T13:30:09.254Z] futures = [<Future: status: cancelled, key: dispatched_train-aeec3bd0-915c-46ba-9098-9dcf8d816571-0>, <Future: status: cancelled, key: dispatched_train-aeec3bd0-915c-46ba-9098-9dcf8d816571-1>]

[2019-10-25T13:30:09.254Z] errors = 'raise', direct = False, local_worker = None

[2019-10-25T13:30:09.254Z] 

[2019-10-25T13:30:09.254Z]     @gen.coroutine

[2019-10-25T13:30:09.254Z]     def _gather(self, futures, errors="raise", direct=None, local_worker=None):

[2019-10-25T13:30:09.254Z]         unpacked, future_set = unpack_remotedata(futures, byte_keys=True)

[2019-10-25T13:30:09.254Z]         keys = [tokey(future.key) for future in future_set]

[2019-10-25T13:30:09.254Z]         bad_data = dict()

[2019-10-25T13:30:09.254Z]         data = {}

[2019-10-25T13:30:09.254Z]     

[2019-10-25T13:30:09.254Z]         if direct is None:

[2019-10-25T13:30:09.254Z]             direct = self.direct_to_workers

[2019-10-25T13:30:09.254Z]         if direct is None:

[2019-10-25T13:30:09.254Z]             try:

[2019-10-25T13:30:09.254Z]                 w = get_worker()

[2019-10-25T13:30:09.254Z]             except Exception:

[2019-10-25T13:30:09.254Z]                 direct = False

[2019-10-25T13:30:09.254Z]             else:

[2019-10-25T13:30:09.254Z]                 if w.scheduler.address == self.scheduler.address:

[2019-10-25T13:30:09.254Z]                     direct = True

[2019-10-25T13:30:09.254Z]     

[2019-10-25T13:30:09.254Z]         @gen.coroutine

[2019-10-25T13:30:09.254Z]         def wait(k):

[2019-10-25T13:30:09.254Z]             """ Want to stop the All(...) early if we find an error """

[2019-10-25T13:30:09.254Z]             st = self.futures[k]

[2019-10-25T13:30:09.254Z]             yield st.wait()

[2019-10-25T13:30:09.254Z]             if st.status != "finished" and errors == "raise":

[2019-10-25T13:30:09.254Z]                 raise AllExit()

[2019-10-25T13:30:09.254Z]     

[2019-10-25T13:30:09.254Z]         while True:

[2019-10-25T13:30:09.254Z]             logger.debug("Waiting on futures to clear before gather")

[2019-10-25T13:30:09.254Z]     

[2019-10-25T13:30:09.254Z]             with ignoring(AllExit):

[2019-10-25T13:30:09.254Z]                 yield All(

[2019-10-25T13:30:09.254Z]                     [wait(key) for key in keys if key in self.futures],

[2019-10-25T13:30:09.254Z]                     quiet_exceptions=AllExit,

[2019-10-25T13:30:09.254Z]                 )

[2019-10-25T13:30:09.254Z]     

[2019-10-25T13:30:09.254Z]             failed = ("error", "cancelled")

[2019-10-25T13:30:09.254Z]     

[2019-10-25T13:30:09.254Z]             exceptions = set()

[2019-10-25T13:30:09.254Z]             bad_keys = set()

[2019-10-25T13:30:09.254Z]             for key in keys:

[2019-10-25T13:30:09.254Z]                 if key not in self.futures or self.futures[key].status in failed:

[2019-10-25T13:30:09.254Z]                     exceptions.add(key)

[2019-10-25T13:30:09.254Z]                     if errors == "raise":

[2019-10-25T13:30:09.254Z]                         try:

[2019-10-25T13:30:09.254Z]                             st = self.futures[key]

[2019-10-25T13:30:09.254Z]                             exception = st.exception

[2019-10-25T13:30:09.254Z]                             traceback = st.traceback

[2019-10-25T13:30:09.254Z]                         except (AttributeError, KeyError):

[2019-10-25T13:30:09.254Z]                             six.reraise(CancelledError, CancelledError(key), None)

[2019-10-25T13:30:09.254Z]                         else:

[2019-10-25T13:30:09.254Z] >                           six.reraise(type(exception), exception, traceback)

[2019-10-25T13:30:09.254Z] 

[2019-10-25T13:30:09.254Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/distributed/client.py:1653: 

[2019-10-25T13:30:09.254Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.254Z] 

[2019-10-25T13:30:09.254Z] tp = <class 'AssertionError'>, value = None, tb = None

[2019-10-25T13:30:09.254Z] 

[2019-10-25T13:30:09.254Z]     def reraise(tp, value, tb=None):

[2019-10-25T13:30:09.254Z]         try:

[2019-10-25T13:30:09.254Z]             if value is None:

[2019-10-25T13:30:09.254Z]                 value = tp()

[2019-10-25T13:30:09.254Z]             if value.__traceback__ is not tb:

[2019-10-25T13:30:09.254Z] >               raise value.with_traceback(tb)

[2019-10-25T13:30:09.254Z] 

[2019-10-25T13:30:09.254Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/six.py:692: 

[2019-10-25T13:30:09.254Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.254Z] 

[2019-10-25T13:30:09.254Z] >   local_mat = mat.get_worker_data(worker)

[2019-10-25T13:30:09.254Z] 

[2019-10-25T13:30:09.255Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/xgboost/dask.py:370: 

[2019-10-25T13:30:09.255Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.255Z] 

[2019-10-25T13:30:09.255Z] >   data, labels, weights = self.get_worker_parts(worker)

[2019-10-25T13:30:09.255Z] 

[2019-10-25T13:30:09.255Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/xgboost/dask.py:266: 

[2019-10-25T13:30:09.255Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2019-10-25T13:30:09.255Z] 

[2019-10-25T13:30:09.255Z] >   assert list_of_parts, 'data in ' + worker.address + ' was moved.'

[2019-10-25T13:30:09.255Z] E   AssertionError: data in tcp://127.0.0.1:46353 was moved.

[2019-10-25T13:30:09.255Z] 

[2019-10-25T13:30:09.255Z] /opt/h2oai/h2o4gpu/python/lib/python3.6/site-packages/xgboost/dask.py:234: AssertionError

[2019-10-25T13:30:09.255Z] ------------------------------ Captured log setup ------------------------------

[2019-10-25T13:30:09.255Z] covtype.py                 110 INFO     Downloading https://ndownloader.figshare.com/files/5976039

[2019-10-25T13:30:09.255Z] ------------------------------ Captured log call -------------------------------

[2019-10-25T13:30:09.255Z] selector_events.py          65 DEBUG    Using selector: EpollSelector

[2019-10-25T13:30:09.255Z] selector_events.py          65 DEBUG    Using selector: EpollSelector

[2019-10-25T13:30:09.255Z] tracker.py                 165 INFO     start listen on 127.0.0.1:9091
```

**Comment by trivialfis** [comment]
Thanks.  Will try to reproduce it tomorrow.

**Comment by trivialfis** [comment]
When the evaluation matrix is too small, some workers don't get a partition.

**Comment by trivialfis** [comment]
One quick workaround for this is:
``` python
dask_test_X = dask_test_X.persist()
client.rebalance(dask_test_X)
```
On real dataset this usually don't happen as one normally load the whole dataset in one file by dask and split it up into multiple train/tests.  The distribution is not changed during the split.

In the long term, we should split up the `train` method into more easily compose-able methods.
